CIP-plutus-modules/README.md

```diff
@@ -0,0 +1,1599 @@
+there is no opportunity to supply additional arguments; all modules
+used must be supplied as `ScriptArg`s, which means they are
+fixed. This makes sense: it would be perverse if a transaction trying
+to spend a UTxO protected by a spending validator were allowed to
+replace some of the validation code--that would open a real can of
+worms, permitting many attacks whenever a script was split over
+several modules. With the design in the CIP, it is the script in the
+UTxO being spent that determines the module versions to be used, not
+the spending transaction. That transaction does need to *supply* all
+the modules actually used--including all of their dependencies--but it
+cannot choose to supply alternative implementations of them.
+
+### In-service upgrade
+
+Long-lived contracts may need upgrades and bug fixes during their
+lifetimes. This is especially true of contracts made up of many
+modules--every time a dependency is upgraded or receives a bug fix,
+the question of whether or not to upgrade the client contract
+arises. However, the problem of upgrading contracts *securely* is a
+tricky one, and exists whether or not modules are used. Therefore this
+CIP does not address this problem specifically: developers should use
+the same mechanisms to handle upgrades of scripts with dependencies,
+as they use to upgrade scripts without dependencies. The only thing we
+note here is that the need for upgrades is more likely to arise when a
+script depends on many others, so it is more important to be prepared
+for it. Note that, because a script contains the hashes of its
+dependencies, no 'silent' upgrade can occur: the hash of a script
+depends on *all* of its code, including the code of its dependencies,
+so any change in a dependency will lead to a change in the script
+hash.
+
+### Lazy loading
+
+The 'lazy loading' variation in the specification section above
+permits fewer modules to be supplied in a transaction.  Dependency
+trees have a tendency to grow very large; when one function in a
+module uses another module, it becomes a dependency of the entire
+module and not just of that function. It is easy to imagine situations
+in which a script depends on many modules, but a particular call
+requires only a few of them. For example, if a script offers a choice
+of protocols for redemption, only one of which is used in a particular
+call, then many modules may not actually be needed. The variation
+allows a transaction to omit the unused modules in such cases. This
+reduces the size of the transaction, which need provide fewer
+witnesses, but more importantly it reduces the amount of code which
+must be loaded from reference UTxOs.
+
+If a script execution *does* try to use a module which was not
+provided, it will encounter a run-time type error and fail (unless the
+module value was `builtin unit`, in which case the script will behave
+as though the module had been provided).

```

Comment:
Is this true?  If modules are allowed to be arbitrary terms then you could supply `let _ = error in ()` and the script would fail, but if that's replaced with `()` then the script might suceed.

I'm not sure that the script would necessarily produce a run-time type error if it tries to use a supposedly unused module which has been replaced by `builtin unit`.  Surely the script could use that module to perform some computation but discard the result of the computation without ever using it.  Maybe that's OK as long as everything is pure, but the presence of the side-effecting `error` in UPLC complicates things.